### PR TITLE
Create gem and add call masking

### PR DESCRIPTION
**Pull request by bt6453**

### CODE REVIEW

**Comment by bt6453** [comment]
Update So Far:

* Added Resource/Model classes for the endpoints we’ll need for call masking/unmasking
* No tests yet. Still got to figure this out.
* Need to get responses on Active Contacts parsed correctly into their `Contact` model
* Still unsure of the approach I came up with. Feels like how we do our normal client gems plus passing  a token around for authentication and trying to make it fit in somehow.

**Comment by bt6453** [comment]
Updates

* Added a central `API` class that accesses each resource and automatically generates a token so we no longer have to pass tokens to every method call (thanks to @a-maas for the suggestion)
* Need to fix the tests

**Comment by bt6453** [comment]
Tests fixed. Going to add a readme and travis-ci integration

File: lib/incontact_client/connection.rb [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    RETRY_SLEEP  = 1
+
+    attr_reader :url
+    attr_reader :authorization
+    attr_reader :default_data_model
+
+    def initialize(url, authorization, options = {})
+      @url                = url
+      @authorization      = authorization
+      @default_data_model = options[:default_data_model] || Hashie::Mash
```
Comment (chmurph2) [inline lib/incontact_client/connection.rb]: Is now our chance to move away from `Hashie::Mash` to POROs, and  maybe `OpenStruct`, now that "lazy" `OpenStruct` is apparently 10x faster in Ruby 2.3? I had some trouble finding concrete evidence that those troubles are behind us, though [this](https://railsisrael2015.events.co.il/presentations/2627-the-value-of-being-lazy-or-how-i-made-openstruct-10x-faster) says so.

File: lib/incontact_client/connection.rb [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    def request_with_exception_handling
+      yield
+    rescue => exception
+      raise ConnectionError, exception
+    end
+
+    def raise_response_errors(response)
+      return if response.success?
+      error_messages = response.body.respond_to?(:exception_message) &&
+        response.body.errors.present? ? response.body.errors.join("\n") : ""
```
Comment (chmurph2) [inline lib/incontact_client/connection.rb]: I want to say this logic often gobbles error messages in our other clients, returning `""` when there is actually an error message.  Just one that isn't in `response.body.exception_message` or `response.body.errors`.  

I think it'd be worth experimenting raising some exceptions to see where this falls short.

File: lib/incontact_client/agent_sessions.rb [hunk 3]
```diff
@@ -0,0 +1,16 @@
+module InContactClient
+  class AgentSessions < TokenAuthenticatedResource
+    def join_agent_session(agent_id)
+      agent_request_data = { "asAgentId" => agent_id }
+      connection.post("services/v8.0/agent-sessions/join", agent_request_data)
```
Comment (bt6453) [inline lib/incontact_client/agent_sessions.rb]: Good call. Will update.

File: lib/incontact_client/agents.rb [hunk 4]
```diff
@@ -0,0 +1,11 @@
+module InContactClient
+  class Agents < TokenAuthenticatedResource
+    def all
+      connection.get("services/v8.0/agents")
+    end
+
+    def find(id)
```
Comment (bt6453) [inline lib/incontact_client/agents.rb]: I actually like `agent_id` here. It’s _technically_ not rails so I’d like to make it more obvious what you’re passing. Plus there’s all kinds IDs floating around.

File: test/test_helper.rb [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+require "dotenv"
+require "webmock/minitest"
+
+Dotenv.load ".env.test", ".env.local", ".env"
+
+module InContactClient
+  class TestCase < MiniTest::Spec
+    TEST_URL = "http://example.com/incontact"
+
+    def mock_request(type, path, status, response_file = nil, request_body = nil, base_url = TEST_URL)
```
Comment (bt6453) [inline test/test_helper.rb]: Will do. I knew once I had 6 params that it was ugly but left it for some unknown reason.

File: lib/incontact_client/api.rb [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+    def agent_sessions
+      @agent_sessions ||=  create_resource(InContactClient::Resources::AGENT_SESSIONS)
+    end
+
+    private
+
+    def create_resource(resource)
+      resource_class = "InContactClient::#{resource}".constantize
+      data_model     = "InContactClient::Responses::#{resource.singularize}Response".constantize
```
Comment (bt6453) [inline lib/incontact_client/api.rb]: Updated

File: lib/incontact_client/connection.rb (hunk 1 above)
Comment (bt6453) [inline lib/incontact_client/connection.rb]: Oooh yeah I’ll look into. There’s a fun little warning that `Hashie::Mash` gives now whenever you try to define a method for a key that already exists and is clogging up the test output. So removing that would be an added bonus.

File: lib/incontact_client/agents.rb (hunk 4 above)
Comment (chmurph2) [inline lib/incontact_client/agents.rb]: 👍

File: test/incontact_client/unit/agent_sessions_test.rb [hunk 7]
```diff
@@ -13,9 +13,10 @@ class AgentSessionsTest < TestCase
 
     describe ".join_agent_session" do
       before do
-        path = "services/v8.0/agent-sessions/join"
+        path = "agent-sessions/join"
         body = { "asAgentId" => 1234 }
-        mock_request(:post, path, 202, "join_session", body)
+        options = { request_body: body, response_file: "join_session", path: path }
+        mock_request(:post, 202, options )
```
Comment (chmurph2) [inline test/incontact_client/unit/agent_sessions_test.rb]: super nitpick: space before `)`

File: README.md [hunk 8]
```diff
@@ -1,2 +1,41 @@
...
+# InContactClient [![Build Status](https://travis-ci.org/customink/incontact_client.svg?branch=master)](https://travis-ci.org/customink/incontact_client)
+
+This gem is a library for accessing [InContact's API](https://developer.incontact.com/API)
+
+
+## Installation
+
+Add this line to your application's Gemfile:
+
+    gem 'shipping_client'
```
Comment (chmurph2) [inline README.md]: Wrong gem name.

File: docs/agent_sessions.md [hunk 9]
```diff
@@ -0,0 +1,30 @@
+### Agent Sessions
+
+You can access agent data using the `InContactClient::Api#agent_session` method.
```
Comment (chmurph2) [inline docs/agent_sessions.md]: "agent" or "agent session" data?

Comment (chmurph2) [inline docs/agent_sessions.md]: Is there a section in the official InContact API docs that's publicly available, that we could link to from these API doc files?

File: README.md (hunk 8 above)
Comment (bt6453) [inline README.md]: SMH that’s what I get for copy/pasting

File: docs/agent_sessions.md (hunk 9 above)
Comment (bt6453) [inline docs/agent_sessions.md]: Agent session data. Thanks. Also I linked to the API in the README.

**Pull request merged**